docs/metadata.rst

```diff
@@ -59,66 +63,190 @@ JSON
-
-    * Endian-ness is fixed at little endian so cannot be specified.
-    * Not all formats are supported - ``binaryFormat`` must match
-      ``^[cbB\?hHiIlLqQfd]|\d*[spx]$``.
-    * It is recommended to use ``p`` for strings, as ``s`` doesn't store where
-      the string ends. Note that ``5p`` can only store a string that when encoded is 4
-      bytes or less. If using ``s``, set ``nullTerminated`` to ``true`` if you are
-      sure your strings do not have null characters in. This will remove the spare
-      bytes at the end of the string.
-    * Strings that are longer than the specified length will be silently truncated,
-      note that the length is in bytes, not characters.
-    * To insert padding bytes use ``null`` for ``type`` and ``x`` for ``binaryFormat``.
-    * For ``array`` types the format of the length of the array can be chosen with
-      ``arrayLengthFormat`` which must be one of ``B``, ``H``, ``I``, ``L`` or ``Q``
-      which is the default. For dealing with legacy encodings that do not store the
-      length of the array setting ``noLengthEncodingExhaustBuffer`` to true will read
-      the buffer to exhaustion. As such an array with this encoding must be the last type
-      in the encoded struct.
-    * For strings, the encoding can be set with ``stringEncoding`` which defaults to
-      ``utf-8``.
+using :py:func:`struct.pack` which results in a compact binary representation which is much smaller
+and generally faster to encode/decode than JSON.
+
+This codec places extra restrictions on the schema:
+
+#. Each property must have a ``binaryFormat``
+    This sets the binary encoding for used for the property.
 
 #. All metadata objects must have fixed properties.
     This means that they can have no missing properties and have no additional
     properties not listed in the schema.
 
 #. Arrays must be lists of homogeneous objects.
-    This is not valid::
+    For example, this is not valid::
 
     {"type": "array", "items": [{"type": "number"}, {"type": "string"}]}
 
 #. Types must be singular and not unions.
-    This is not valid::
+    For example, this is not valid::
 
     {"type": ["number", "string"]}
 
+    One exception is that the top-level can be a union of ``object`` and ``null`` to
+    support the case where some rows do not have metadata.
+
 #. The order that properties are encoded is by default alphabetically by name.
     The order can be overridden by setting an optional numerical ``index`` on each
-    property.
-
+    property. This is due to objects being unordered in JSON and Python ``dicts``.
+
+++++++++++++
+binaryFormat
+++++++++++++
+
+To determine the binary encoding of each property in the metadata the ``binaryFormat`` key is used.
+This describes the encoding for each property using ``struct``
+`format characters <https://docs.python.org/3/library/struct.html#format-characters>`_.
+For example an unsigned 8-byte integer can be specified with::
```

Comment:
We should probably acknowledge some of the below as being modified from the struct docs?

Reply:
Fixed in https://github.com/tskit-dev/tskit/pull/741/commits/3d63b4994192e631b1721f128050bfeeba7ce509